### PR TITLE
disabling FI tracking by deactivating pixel

### PR DESCRIPTION
**Pull request by luguenth**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | staging
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | no
| Related user documentation PR URL      | n.A.
| Related developer documentation PR URL | n.A.
| Issue(s) addressed                     | n.A.

#### Description:

We wanted to get rid of tracking when using Focus Items. So we made a setting for the FocusBundle to not include the Tracking-Pixel if disabled. So there'll be no more Tracking-Cookies any more.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Go to the Configuration Page
3. Select Focus Settings.
4. Disable the Tracking-Pixel
5. Create a FI
6. embed this item at another Website
7. Look into the DevConsole, there shouldn't be any mtc-tracking cookies anymore. Except the FI-Cookie

### CODE REVIEW

**Comment by ekkeguembel** [comment]
Background: This allows Focus Items to be used even if a **Consent Manager** is used and the user has denied consent to tracking, marketing cookies etc.
The mautic_focus cookie will still be set (can be defined as "technically required") but nothing else.

File: plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php]:
```suggestion
 * @copyright   2020 Mautic Contributors. All rights reserved
```

File: plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+use Mautic\ConfigBundle\ConfigEvents;
+use Mautic\ConfigBundle\Event\ConfigBuilderEvent;
+use Mautic\ConfigBundle\Event\ConfigEvent;
+use Mautic\PageBundle\Form\Type\ConfigTrackingPageType;
+use Mautic\PageBundle\Form\Type\ConfigType;
+use MauticPlugin\MauticFocusBundle\Form\Type\FocusConfigTrackingPageType;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class ConfigSubscriber implements EventSubscriberInterface
```
Comment (escopecz) [inline plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php]: Some offensive programming won't hurt.
```suggestion
final class ConfigSubscriber implements EventSubscriberInterface
```

File: plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /**
+     * @return array
+     */
+    public static function getSubscribedEvents()
```
Comment (escopecz) [inline plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php]: The method comment can be removed when we can use return types for new methods in Mautic 3 🎉 
```suggestion
    public static function getSubscribedEvents(): array
```

File: plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return [
+            ConfigEvents::CONFIG_ON_GENERATE => ['onConfigGenerateTracking', 0],
+        ];
+    }
+
+    public function onConfigGenerateTracking(ConfigBuilderEvent $event)
```
Comment (escopecz) [inline plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php]:
```suggestion
    public function onConfigGenerateTracking(ConfigBuilderEvent $event): void
```

File: plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php]:
```suggestion
 * @copyright   2020 Mautic Contributors. All rights reserved
```

File: plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
+<?php
+
```
Comment (escopecz) [inline plugins/MauticFocusBundle/EventListener/ConfigSubscriber.php:2]:
```suggestion

declare(strict_types = 1);
```

File: plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
+<?php
+
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php:2]:
```suggestion

declare(strict_types = 1);
```

File: plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+namespace MauticPlugin\MauticFocusBundle\Form\Type;
+
+use Mautic\CoreBundle\Form\Type\YesNoButtonGroupType;
+use Mautic\PageBundle\Form\Type\ConfigTrackingPageType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+/**
+ * Class ConfigType.
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php]: Please remove this class comment completely.

File: plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        $builder->add(
+            'focus_pixel_enabled',
+            YesNoButtonGroupType::class,
+            [
+                'label' => 'mautic.page.config.form.focus.pixel.enabled',
+                'attr'  => [
+                    'tooltip' => 'mautic.page.config.form.focus.pixel.enabled.tooltip',
+                ],
+                'data'  => isset($options['data']['focus_pixel_enabled']) ? (bool) $options['data']['focus_pixel_enabled'] : false,
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Form/Type/FocusConfigTrackingPageType.php]: Why is the default value false?
```suggestion
                'data'  => (bool) $options['data']['focus_pixel_enabled']) ?? true,
```

File: plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            ->method('add')
+            ->with(
+                'focus_pixel_enabled',
+                YesNoButtonGroupType::class,
+                [
+                    'label' => 'mautic.page.config.form.focus.pixel.enabled',
+                    'attr'  => [
+                        'tooltip' => 'mautic.page.config.form.focus.pixel.enabled.tooltip',
+                    ],
+                    'data'  => isset($options['data']['focus_pixel_enabled']) ? (bool) $options['data']['focus_pixel_enabled'] : false,
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php]: Why is the default value false?
```suggestion
                'data'  => (bool) $options['data']['focus_pixel_enabled']) ?? true,
```

File: plugins/MauticFocusBundle/Tests/EventListener/ConfigSubscriberTest.php [hunk 11]
```diff
@@ -0,0 +1,47 @@
+<?php
+
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/EventListener/ConfigSubscriberTest.php:2]:
```suggestion

declare(strict_types = 1);
```

File: plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php [hunk 12]
```diff
@@ -0,0 +1,63 @@
+<?php
+/*
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php:5]:
```suggestion

declare(strict_types = 1);

/*
```

File: plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+                        'tooltip' => 'mautic.page.config.form.focus.pixel.enabled.tooltip',
+                    ],
+                    'data'  => isset($options['data']['focus_pixel_enabled']) ? (bool) $options['data']['focus_pixel_enabled'] : false,
+                ]
+            );
+
+        $this->form->buildForm($this->formBuilder, $options);
+    }
+
+    public function testGetBlockPrefix()
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php]:
```suggestion
    public function testGetBlockPrefix(): void
```

File: plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    private $formBuilder;
+
+    public function setUp()
+    {
+        $this->form                 = new FocusConfigTrackingPageType();
+        $this->formBuilder          = $this->createMock(FormBuilderInterface::class);
+    }
+
+    public function testBuildForm()
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php]:
```suggestion
    public function testBuildForm(): void
```

File: plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+     * @var FocusConfigTrackingPageType
+     */
+    private $form;
+
+    /**
+     * @var \PHPUnit\Framework\MockObject\MockObject
+     */
+    private $formBuilder;
+
+    public function setUp()
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/Form/Type/FocusConfigTrackingPageTypeTest.php]:
```suggestion
    public function setUp(): void
```

File: plugins/MauticFocusBundle/Tests/EventListener/ConfigSubscriberTest.php [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+     * @var ConfigEvent|MockObject
+     */
+    private $configEvent;
+
+    /**
+     * @var ConfigSubscriber
+     */
+    private $subscriber;
+
+    protected function setUp()
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/EventListener/ConfigSubscriberTest.php]:
```suggestion
    protected function setUp(): void
```

File: plugins/MauticFocusBundle/Tests/EventListener/ConfigSubscriberTest.php [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        $this->configEvent          = $this->createMock(ConfigEvent::class);
+        $this->subscriber           = new ConfigSubscriber();
+    }
+
+    public function testGetSubscribedEvents()
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Tests/EventListener/ConfigSubscriberTest.php]:
```suggestion
    public function testGetSubscribedEvents(): void
```

File: plugins/MauticFocusBundle/Translations/en_US/messages.ini [hunk 18]
```diff
@@ -109,3 +109,7 @@ mautic.campaign.focus.show="Focus Item"
 mautic.focus.form.confirmdelete="Delete the Focus item, %name%?"
 mautic.focus.protocol.mismatch="This site cannot be previewed because it is not secure. Please try using %url%."
 mautic.focus.blocking.iframe.header="This website blocks iframe previews with the %header% header. Please try another URL."
+mautic.config.tab.focusconfig = "Focus Settings"
+mautic.page.config.form.focus.pixel.enabled = "Enable Tracking-Pixel for Focus Items"
+mautic.page.config.form.focus.pixel = "Focus Item"
+mautic.page.config.form.focus.pixel.enabled.tooltip = "If set to no, there will be no Tracking Cookies set."
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Translations/en_US/messages.ini]: Let's add the reason for why would anyone want to turn it off to the tooltip.
```suggestion
mautic.page.config.form.focus.pixel.enabled.tooltip = "If set to no, there will be no Tracking Cookies set. This allows Focus Items to be used even if a Consent Manager is used and the user has denied consent to tracking, marketing cookies etc. The mautic_focus cookie will still be set (can be defined as "technically required") but nothing else."
```

File: plugins/MauticFocusBundle/Model/FocusModel.php [hunk 19]
```diff
@@ -260,10 +260,11 @@ public function getContent(array $focus, $isPreview = false, $url = '#')
...
             [
-                'focus'    => $focus,
-                'preview'  => $isPreview,
-                'htmlMode' => $htmlMode,
-                'clickUrl' => $url,
+                'focus'            => $focus,
+                'preview'          => $isPreview,
+                'htmlMode'         => $htmlMode,
+                'clickUrl'         => $url,
+                'tracking_enabled' => $this->coreParametersHelper->get('focus_pixel_enabled'),
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Model/FocusModel.php]:
```suggestion
                'tracking_enabled' => $this->coreParametersHelper->get('focus_pixel_enabled', true),
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Great feature! I've found a couple ways how to improve the code. Please check if it makes sense for you. 

Also the CodeCov check is failing. Please add one more test. Check the CodeCov details to see what can be still covered.

**Comment by luguenth** [comment]
Did I overlook some change requests? They seem resolved for me. But GitHub still states there are requests, where can I find them?

**Comment by luguenth** [comment]
@escopecz can you recheck the PR? There still seem to be requested changes, but I can't find them.

File: plugins/MauticFocusBundle/Translations/en_US/messages.ini [hunk 20]
```diff
@@ -109,3 +109,7 @@ mautic.campaign.focus.show="Focus Item"
 mautic.focus.form.confirmdelete="Delete the Focus item, %name%?"
 mautic.focus.protocol.mismatch="This site cannot be previewed because it is not secure. Please try using %url%."
 mautic.focus.blocking.iframe.header="This website blocks iframe previews with the %header% header. Please try another URL."
+mautic.config.tab.focusconfig = "Focus Settings"
+mautic.page.config.form.focus.pixel.enabled = "Enable Tracking-Pixel for Focus Items"
+mautic.page.config.form.focus.pixel = "Focus Item"
+mautic.page.config.form.focus.pixel.enabled.tooltip = "If set to no, there will be no Tracking Cookies set. This allows Focus Items to be used even if a Consent Manager is used and the user has denied consent to tracking, marketing cookies etc. The mautic_focus cookie will still be set (can be defined as "technically required") but nothing else."
```
Comment (escopecz) [inline plugins/MauticFocusBundle/Translations/en_US/messages.ini]: The inner double quotes will have to be escaped or changed to single quotes.

```suggestion
mautic.page.config.form.focus.pixel.enabled.tooltip = "If set to no, there will be no Tracking Cookies set. This allows Focus Items to be used even if a Consent Manager is used and the user has denied consent to tracking, marketing cookies etc. The mautic_focus cookie will still be set (can be defined as \"technically required\") but nothing else."
```

**Comment by luguenth** [comment]
Alright, I accepted your Code Changes :) Could you re-review please? :rocket:

**Comment by escopecz** [comment]
I see no issues in the code. But I'm a bit confused from the testing. I see no cookies stored at all even without this configuration. I can see `mautic_device_id`, `mtc_sid` and `mtc_id` keys and values stored in local storage. And this happens when I enable or disable this feature. Am I missing something?

**Comment by RCheesley** [comment]
@luguenth any tips to help @escopecz with reproducing the issue this PR is meant to be fixing?

**Comment by ekkeguembel** [comment]
@RCheesley @escopecz we reviewed today, and decided that we are unsure about the exact _desired_ behavior in conjunction with various consent managers.
-> Let's not put this into 4.0 but rather take a bit of time on Monday and make sure we do the right thing.

**Pull request closed without merging**